### ISSUE TITLE
Reconstruction and Editability benchmark for experiments 

### ISSUE THREAD

**Issue by SoyGema** [rinongal/textual_inversion#102]
Hello there! 
Thanks for providing a low level evaluation for benchmarking the quality of the latent space embeddings . 👍 💯 
I have some questions regarding **Reconstruction** and Editability concepts:

1 ) **Reconstruction** . 
From the paper I´m understanding that the metric used as a quality benchmark for the reconstructed concept is the average PAIR CLIP-space cosine-similarity between the generated images and the images of the concept-specific training set .It takes 64 images generated using the prompt “A photo of S*” .

Therefore, the closest to 1 the better the reconstruction might have been produced. Is this vector thought correct?

- For computing this metric with different configs, I shall produce 64 images with “A photo of S*” and the **3-5 training images** and then compute the average PAIR CLIP-space cosine-similarity in between them. Is this correct ?

- Is this metric implemented somewhere in the code and can it be printed easily ?

2 ) **Editability**
From here, what Im understanding is that you measure CLIP-space embedding of the samples, and compute their cosine similarity with the CLIP-space embedding of the textual prompts omiting S* . So , as I´m understanding, we take a generated image with S* , and then compare it with the textual prompt without S* , is this correct ? Do you know if using [this](https://huggingface.co/spaces/h4d35/CosineSim) might work ? 

_Last , but not least, could we say that the difference in between those metrics is that in Reconstruction we measured similarity in between images and in Editability we measure similairty in between the image and the text prompt ?_

Thanks so much and congratulations again for the fanstastic work ! Super Inspiring!

**Comment by rinongal** [comment]
Hi,

First of all, regarding implementations of the evaluation script, you can find the relevant code to run similar evaluations [here](https://github.com/rinongal/textual_inversion/blob/main/evaluation/clip_eval.py) and [here](https://github.com/rinongal/textual_inversion/blob/main/scripts/evaluate_model.py). I just ran the later file with a batch script, but if you want that code as well let me know and I'll tidy it up asap.

It does all the clip calculations including cosine similarity etc, so you don't need to run that gradio demo manually.

On the rest of the questions: 

- Keep in mind the string that you should actually be using to test reconstruction is "a photo of *". We used S* for better clarity in the paper, but the actual default placeholder string in our code is just '*'.

- You are correct on the interpretation of the metrics. Reconstruction checks the similarity between generated images and the original set, and editability checks similarity between the generated images and the text prompts.

**Comment by SoyGema** [comment]
Thanks for the detailed help and rigourousness towards the answer.
Following the trade-off exploration in between editability and reconstruction coming from section 5.3 Results , in page 13 , 1st paragraph (just after figure) , I´m assuming 

-  the closer the learned embedding coming from * and the lower the learning rate, the better the **editability** is, but it shall fail to catch concept details. In contrast, the deviation gives a better **reconstruction** but a lower editability. The metric that controls this tradeoff is the learning rate. Could you please confirm when you have the time that we are taking about the [base_learning_rate](https://github.com/rinongal/textual_inversion/blob/0a950b482d2e8f215122805d4c5901bdb4a6947f/configs/stable-diffusion/v1-finetune.yaml#L2) coming from the model  yaml ?

- Have you found a learning rate / range interesting to be benchmarked that could be consider **robust enough** for extended latent spaces ?

- Tangential : Fully aware that this vector thought deviates from the conversation and this is just an idea for future benchmarking and metrics (could be wrong, take this with a pinch of salt...). For evaluation ,at some point [AlphaStar](https://openreview.net/forum?id=Np8Pumfoty) supplementary material A.5- introduced a evaluation _robustness_ metric .
Maybe in this case a _robust *_  would be a probabilty distribution closest to 1 in between the CLIP quotients coming from R/E could gives us an idea of how well the concept has been learnt and is implemented in a context ?
Thanks ! 👍

**Comment by SoyGema** [comment]
Hello there,
My apologies for the follow up in another comment. I think it might be more fruitfull to touch Tangential discussions once  supplementary material is digested on my side. Thanks for your support again!

**Comment by rinongal** [comment]
Hi,

On the question of LR controlling the tradeoff - the paper refers to the base_learning_rate, yes. However, please note that this isn't the only way to control the tradeoff. Adding more tokens, or adding regularizations to keep the embedding close to real words can also have a similar effect.

The learning rate we ended out settling on is the repo's default of 5e-03. It is near the 'corner' of the tradeoff curve.

We did not consider robustness scores, but if you look at our openreview supplementary you'll see a new experiment about coverage which is semi-related.

**Comment by rinongal** [comment]
Learning rate:
The unfrozen yaml has two different learning rates, both of which are being used. 
The first is the base_learning_rate, which is used as before to optimize the model embedding. 
The second is the model_lr, which is used when optimizing the model's parameters. Basically this lets you run the pivotal tuning experiment described in our paper, and use a different learning rate for each part.

Additional approaches:
Bipartite and Pivotal are indeed two more advanced methods for inversion, and they are not presented in the main paper mostly over length concerns.

I'm not sure I follow the question about scale factor. It is set to the same value for both the frozen and the unfrozen versions of the model. Keep in mind that the scale factor in the config is **not** the classifier guidance scale. You're supplying the classifier guidance scale when running inference with the model.

We did not play with the scale_factor param in the config, we're just using CompViz' defaults.

Regularization:
Yes. What we basically checked was whether staying closer to the distribution of real words would help with editability, and this is indeed the case. It did hurt identity preservation, however.

I can confirm that the parameter you pointed at in the yaml is controlling the strength of this regularization. You want to keep it fairly low (probably around 1e-6 though your mileage may vary if you're using Stable Diffusion and not LDM).

**Comment by SoyGema** [comment]
Hello @rinongal ! Thanks so much for the detailed answer. 

**Learning Rate**

Thanks so much for the clarification of the learning rates use.  


**Scale in Additional approaches VS scale guidance during inference** 

In the additional approaches, the image matrix shows different scales ( s =1, s=2, s=5 ) . I was wondering if conceptually speaking it was the same as classifier guidance scale during inference.  Now I see I had a missconception. Thanks for pointing me out in the right direction . 
Regarding this **guidance_scale_factor** , Im assuming do you recommend Iterating over this in inference for controlling the tradeoff ?  

I think Im having a good enough perspective and I don´t want to abuse your kindness , so If you want we can agree on discussing over this last question and close the issue. 
[We](https://github.com/0x2b3bfa0) are working with all this to release a case of use with [CI/CD ](https://github.com/iterative/cml) on top of your adsorbing and refreshing work, and more than happy to amplify the voice of your research . 
Hopefully you don´t mind we thank you in one of our articles for all your support 🥇

**Comment by rinongal** [comment]
Sorry for the late response. Busy week :)

For guidance_scale_factor you probably want to keep it somewhere near the original values (so 7.0-10.0). Higher will likely give you saturation issues without much benefit to the prompt-matching, low values will probably create things unrelated to your prompt.

Don't worry about closing the issue, you're more than welcome to ask questions, that's what this place is for.
